### PR TITLE
[RSDK-1054] Fix delete_processed_data bug

### PR DESCRIPTION
**Pull request by kkufieta**
Ticket: https://viam.atlassian.net/browse/RSDK-1621

Integration tests will pass once this is merged: https://github.com/viamrobotics/rdk/pull/1726

Tested manually:
* Cartographer on osx
* Orbslam on rpi

[RSDK-1621]: https://viam.atlassian.net/browse/RSDK-1621?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: slam-libraries/viam-cartographer/src/slam_service/slam_service.cc [hunk 1]
```diff
@@ -451,6 +451,12 @@ std::string SLAMServiceImpl::GetNextDataFileOnline() {
     while (b_continue_session) {
         const auto file_list_online =
             viam::io::ListSortedFilesInDirectory(path_to_data);
+        if (delete_processed_data && first_processed_file_index >= 0) {
+            for (int i = first_processed_file_index;
```
Comment (jeremyhyde-viam) [inline slam-libraries/viam-cartographer/src/slam_service/slam_service.cc:455]: [q] Correct me if I'm wrong but `first_processed_file_index` is being used to ensure that we delete all files AFTER the one that was previously found by `GetNextDataFile`, including said file.

Shouldn't this be all the files before this one? Or is that to protect previous runs of data?

File: slam-libraries/viam-cartographer/src/slam_service/slam_service.h [hunk 2]
```diff
@@ -111,6 +111,10 @@ class SLAMServiceImpl final : public SLAMService::Service {
     std::atomic<bool> optimize_on_start{false};
     std::atomic<bool> offline_flag{false};
     bool delete_processed_data = false;
+    // The size of the buffer has to be the same as
+    // dataBufferSize in RDK's builtin_test.go
+    const int data_buffer_size = 4;
+    int first_processed_file_index = -1;
```
Comment (jeremyhyde-viam) [inline slam-libraries/viam-cartographer/src/slam_service/slam_service.h:117]: [q] Does this need to be a global?

File: slam-libraries/viam-orb-slam3/orbslam_server_v1.h [hunk 3]
```diff
@@ -60,6 +60,10 @@ class SLAMServiceImpl final : public SLAMService::Service {
     double yamlTime;
     std::atomic<bool> offlineFlag{false};
     bool delete_processed_data = false;
+    // The size of the buffer has to be the same as
+    // dataBufferSize in RDK's builtin_test.go
+    const int data_buffer_size = 4;
+    int first_processed_file_index = -1;
```
Comment (jeremyhyde-viam) [inline slam-libraries/viam-orb-slam3/orbslam_server_v1.h:66]: see previous question

**Review by jeremyhyde-viam — CHANGES_REQUESTED** [review]
A few questions before I sign off

File: slam-libraries/viam-cartographer/src/slam_service/slam_service.cc [hunk 4]
```diff
@@ -565,6 +571,16 @@ void SLAMServiceImpl::ProcessDataAndStartSavingMaps(double data_start_time) {
...
             }
+            // Get index of the first file we're reading in
+            const auto files =
+                viam::io::ListSortedFilesInDirectory(path_to_data);
+            auto file_index = std::find(begin(files), end(files), file);
+            if (file_index == std::end(files)) {
+                throw std::runtime_error(
+                    "the file should be in the list of files: " + file);
+            }
+            first_processed_file_index =
```
[the remark this answers is not in the document]
Comment (JohnN193) [reply slam-libraries/viam-cartographer/src/slam_service/slam_service.cc:582]: I think this value is non-zero if we already had data in our directory

**Review by JohnN193 — APPROVED** [review]
LGTM

File: slam-libraries/viam-cartographer/src/slam_service/slam_service.cc (hunk 1 above)
Comment (kkufieta) [reply slam-libraries/viam-cartographer/src/slam_service/slam_service.cc:455]: This is to protect previous runs of data.

Comment (jeremyhyde-viam) [reply slam-libraries/viam-cartographer/src/slam_service/slam_service.cc:455]: I'd argue that ListSortedFilesInDirectory should only return current runs of data

Comment (kkufieta) [reply slam-libraries/viam-cartographer/src/slam_service/slam_service.cc:455]: I like this idea - or at least a variation of it. I'll try that out.

Comment (kkufieta) [reply slam-libraries/viam-cartographer/src/slam_service/slam_service.cc:455]: Created a ticket to fix this: https://viam.atlassian.net/browse/RSDK-1659

Reason is that I want this bug fix to go in for the rest of the team.

**Review by jeremyhyde-viam — APPROVED** [review]
LGTM with my unresolved comment being made into a separate JIRA ticket

**Pull request merged**